src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -207,33 +290,60 @@ public void pair(APDU apdu) {
   }
 
   /**
-   * Performs the first step of pairing. In this step the card solves the challenge sent by the client, thus authenticating
-   * itself to the client. At the same time, it creates a challenge for the client. This can only fail if the card has
-   * already paired with the maximum allowed amount of clients.
+   * Performs the first step of certificate based pairing. In this step, the card provides a CA signed certificate
+   * of its card ID key, and proves ownership of this key with a signature on a challenge hash. The challenge hash
+   * to be signed is computed as the sha256 hash of the client salt (provided by client) and card salt (generated randomly).
+   * The card will also include its random salt in the response to the client, so that the client can reproduce the challenge
+   * hash and verify the card signature.
    *
    * @param apduBuffer the APDU buffer
    * @return the length of the reply
    */
   private short pairStep1(byte[] apduBuffer) {
-    preassignedPairingOffset = -1;
-
-    for (short i = 0; i < (short) pairingKeys.length; i += PAIRING_KEY_LENGTH) {
-      if (pairingKeys[i] == 0) {
-        preassignedPairingOffset = i;
-        break;
-      }
+    // Make sure certificate exisits
+    if (idCertStatus != ID_CERTIFICATE_LOCKED) {
+      ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
     }
 
-    if (preassignedPairingOffset == -1) {
-      ISOException.throwIt(ISO7816.SW_FILE_FULL);
+    // Override first pairing slot
+    preassignedPairingOffset = 0;
+
+    // Compute ECDH secret
+    crypto.ecdh.init(idKeypair.getPrivate());
+    short pubKeyOff = (short)(ISO7816.OFFSET_CDATA + SC_SECRET_LENGTH);
+    try {
+      crypto.ecdh.generateSecret(apduBuffer, (short) (pubKeyOff + 1), apduBuffer[pubKeyOff], secret, (short) 0);
+    } catch(Exception e) {
+      ISOException.throwIt(ISO7816.SW_WRONG_DATA);
+      return 0;
     }
 
-    crypto.sha256.update(pairingSecret, (short) 0, SC_SECRET_LENGTH);
-    crypto.sha256.doFinal(apduBuffer, ISO7816.OFFSET_CDATA, SC_SECRET_LENGTH, apduBuffer, (short) 0);
-    crypto.random.generateData(secret, (short) 0, SC_SECRET_LENGTH);
-    Util.arrayCopyNonAtomic(secret, (short) 0, apduBuffer, SC_SECRET_LENGTH, SC_SECRET_LENGTH);
+    // Hash client salt and ECDH secret and store in secret buffer
+    // secretHash = sha256(clientSalt, ECDH secret)
```

Comment:
Guessing you don't need this cruft?

Reply:
That was my attempt to document how `secretHash` is constructed, but it does kind of look like cruft. I can remove if its not helpful